lib/resque/scheduler.rb

```diff
@@ -48,38 +56,46 @@ def run
         $stdout.sync = true
         $stderr.sync = true
 
-        was_master = nil
+        @was_master = nil
 
         begin
           @th = Thread.current
 
           # Now start the scheduling part of the loop.
-          loop do
-            begin
-              # Check on changes to master/child
-              @am_master = master?
-              if am_master != was_master
-                procline am_master ? 'Master scheduler' : 'Child scheduler'
-
-                # Load schedule because changed
-                reload_schedule!
-              end
+          loop { main_loop_once }
 
-              if am_master
-                handle_delayed_items
-                update_schedule if dynamic
-              end
-              was_master = am_master
-            rescue *INTERMITTENT_ERRORS => e
-              log! e.message
-              release_master_lock
+        rescue Interrupt
+          log 'Exiting'
+        end
+      end
+
+      def main_loop_once
+        begin
+          # Check on changes to master/child
+          @am_master = master?
+          if am_master != @was_master
+            procline am_master ? 'Master scheduler' : 'Child scheduler'
+
+            reload_schedule!
+            if am_master
+              log!("Assuming master role.")
+              # It's a protected method for some reason.
+              rufus_scheduler.__send__(:start)
+            else
+              log!("Assuming non-master role, not running cron jobs.")
+              @rufus_scheduler&.stop(:wait)
             end
-            poll_sleep
           end
 
-        rescue Interrupt
-          log 'Exiting'
+          handle_delayed_items
```

Comment:
I'm going to tag a few spots here that look like major behavioural changes, mostly as notes to myself, but also maybe as places for clarification:

It appears we used to handle delayed items only if master, but now this happens for any scheduling process?

Reply:
Correct. Previously only the master would process delayed items, but now all schedulers can. This is the main purpose of this PR.